### PR TITLE
Allow writing null valued keys in JSON

### LINKED ISSUE #3774: arrow_json::LineDelimitedWriter: Write nulls
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I am trying to convert Parquet data written using Avro Parquet Writer of Java. This parquet includes the Avro Schema inside the key `parquet.avro.schema` of the metadata, I want to convert this parquet data back to Avro using this schema and programmed in Rust.

I have tried to solve this challenge by converting Parquet into a json with string `arrow_json::LineDelimitedWriter` and then read this data with `serde_json` and serialise to Avro. In Avro Serializer I need that nullable fields are provided as `Option<T>`. The problem is that when I write the data into JSON null fields are skipped and I can't recover them as Options with `serde_json`

**Describe the solution you'd like**

I would like that `arrow_json` has a parameter for specifying if you want to output the null field.

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #3774

# Rationale for this change
 

Allow users to write JSON with null values for keys, where previous (and default) is to skip the keys.

# What changes are included in this PR?

In arrow-json writer mod, new builder for `Writer` which allows configuring how null valued keys (and null arrays) should be handled. Will default to skipping keys with null values for backward compatibility.

# Are there any user-facing changes?

Yes

### CODE REVIEW

File: arrow-json/src/writer.rs [hunk 1]
```diff
@@ -494,21 +575,36 @@ impl JsonFormat for JsonArray {
...
+/// type parameter. The [`NullHandler`] type parameter controls whether
+/// nulls should be written explicitly for keys or skipped. Default is
+/// [`SkipNulls`] for backward compatibility.
 #[derive(Debug)]
-pub struct Writer<W, F>
+pub struct Writer<W, F, N = SkipNulls>
 where
     W: Write,
     F: JsonFormat,
+    N: NullHandler,
```
Comment (Jefffrey) [inline arrow-json/src/writer.rs:607]: I wasn't sure if it would be better to include this null handling logic in `JsonFormat` instead. Open to suggestions

**Comment by tustvold** [comment]
As commented on the ticket, the current behaviour is simply incorrect as it will not survive a round-trip. I therefore don't think there is a need to make the prior logic configurable, which I think should simplify this PR

**Comment by Jefffrey** [comment]
> As commented on the ticket, the current behaviour is simply incorrect as it will not survive a round-trip. I therefore don't think there is a need to make the prior logic configurable, which I think should simplify this PR

If you're referring to #5066 then that is a separate issue which this PR doesn't cover.

**Comment by tustvold** [comment]
Aah I see, in that case is there a way we could make this an option instead of a generic. As it stands this is a breaking change, which I'm not sure about

**Comment by Jefffrey** [comment]
Can you give an example where this might break existing uses? I thought including a default for the new generic parameter for `Writer` would make this backwards compatible,  but would be happy to be proven wrong.

**Comment by Jefffrey** [comment]
> Regardless using generics here will result in additional codegen and I'm doubtful will improve performance. Adding a WriterOptions or similar should be more future proof and more consistent with other implementations

Yeah that makes 👍

I'll work on switching to such

**Comment by Jefffrey** [comment]
Removed configuring this null behaviour as a generic trait, and changed to be a config that can be set via a new WriterBuilder (akin to arrow-csv writing behaviour)

File: arrow-json/src/writer.rs (hunk 1 above)
Comment (Jefffrey) [reply arrow-json/src/writer.rs]: Changed to be configured via a WriterBuilder now

**Review by tustvold — APPROVED** [review]
This looks good to me, my only comment is I think it should be `keep_null_values` or possibly `preserve_nulls`, as null keys are actually illegal in JSON :sweat_smile:

**Comment by Jefffrey** [comment]
> This looks good to me, my only comment is I think it should be `keep_null_values` or possibly `preserve_nulls`, as null keys are actually illegal in JSON 😅

Yeah I wasn't sure on the proper name. I was gonna leave it as `keep_nulls` but I felt that could be confusing since it might imply affecting null values in arrays which should always be kept regardless.

Maybe something like `preserve_object_nulls`?

**Comment by tustvold** [comment]
Perhaps explicit_nulls ?

**Comment by Jefffrey** [comment]
> Perhaps explicit_nulls ?

Sounds better, renamed to that :+1:

**Review by tustvold — APPROVED** [review]
Looks good to me, thank you

**Pull request merged**